tools/python/runfiles/runfiles.py

```diff
@@ -114,6 +117,7 @@ class _Runfiles(object):
 
   def __init__(self, strategy):
     self._strategy = strategy
+    self._python_runfiles_root = _FindPythonRunfilesRoot()
```

Comment:
Have the caller of the _Runfiles object pass the runfiles directory in as an argument. This avoids most of the issues with trying to re-find the runfiles directory.

In CreateDirectoryBased(), the path can be directly passed in.

In CreateManifestBased(), you can use the same logic the launcher itself uses, which is pretty simple:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt;l=134-144;drc=c3425feeb3bc204923979773ae985dd2f3e24b9f

Reply:
Doing that leads to this test failure: https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/0183d34f-4545-4f97-933c-3164ea7a67b1/tools%5Cpython%5Crunfiles%5Crunfiles_test%5Ctest.log
Do you know whether there is a convenient way to handle UNC paths in Python?

Reply:
Hm. I think you can simply strip of `\\?` and `\\.`. From the docs I've found, it looks like those are special "dos device paths" that refer to the current device, and the difference between `\\?` and `\\.` is something to do how windows does path normalization.

Or maybe check the manifest mapping for the caller's file? _ManifestBased._result has a `runfile_path -> fs path` mapping. If it can be found in there, it's not a big deal to maintain a reverse map.

Reply:
Did that, but this is now hitting a more fundamental problem (see [these logs](https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/0183d5ab-b35e-4690-87e1-93fbdd0894f5/src%5Ctest%5Cpy%5Cbazel%5Cpy_test%5Ctest.log)): The Bazel runfiles root may very well be distinct from the Python runfiles root when using `--build_python_zip`. But the lookup of the current repository should always use the Python runfiles root.

The previous approach handled this distinction. Would you be okay with reverting to it?

Reply:
Upfront: Thanks for being diligent and persisting through the review of this arguably weird PR! My situation is really just that we have to make runfiles work as well as possible with the drastic changes that Bzlmod necessarily brings for them. I know this is certainly not ideal and gets us into the fringe parts of the different languages - you don't want to know how we are planning to implement `CurrentRepository` in Java ;-)

> How about just looking for ".runfiles" (or the other substrings for the other various cases) in the caller's file path?
> 
> I forget what the previous approach was. The module-level _FindRunfilesRoot() function, based upon sys.argv? I could accept that at this point. Not using sys.path, though.

I preserved it in the first commit: https://github.com/bazelbuild/bazel/pull/16341/commits/4b870459f32b93177184446aed74705e1f2dccac#diff-21d70609ca9dc507b2d33f4a68d89bca3cd727191ffe3a6319b63c36e750deb2R215

It's quite similar to scanning for `.runfiles` in the scripts own file path, but slightly more precise since we just know the suffix we need to strip.

> That said, part of the point I'm trying to make here -- one that I'm not really convinced is landing -- is that there just isn't any good way to get the runfiles root after program startup. sys.argv isn't reliable because user code may change it, or it may point to a symlink that points elsewhere. Looking at the callstack is iffy because not all callers have a file, and a caller doesn't necessarily have to live in the runfiles. If we look at all the code surrounding "where is the runfiles root?" (i.e. here and the stub launcher), we see it's very non-trivial with many special cases.

I do think that it is important to consider the following two aspects of this PR separately. Doing that will also help me understand your point better:

1. Finding the Python runfiles root:
I don't yet see why this should be difficult: The script at `bazel_tools/tools/py/runfiles/runfiles.py` is an honest Python file and it lies under the runfiles root because the launcher either found the Bazel runfiles directory with this file in the known relative location or set up a runfiles-like directory by extracting the ZIP. Deriving the root from that requires neither `sys.argv` nor `sys.path`, we just strip a suffix from this one particular file's path. 
This is important because if this part of the runfiles library isn't reliable, the library can't really be used at all.
 
2. Using that information to derive the Bazel repository containing the caller:
This is where we invoke a lot of magic that may break in a myriad of ways: It requires investigating the call stack, assumes that the caller is a file, assumes that it lies in the runfiles root, and so on. 
Luckily for us, if this magic procedure breaks from a particular location, it's not too bad: It just means that users can't look up their Bazel context *from that particular location*. Assuming they want to use runfiles, they will likely find some more well-behaved place in their program where the lookup is successful and then pass the repo name on to code that can then call `Rlocation` without any further magic involved.

Does this assessment agree with what you are thinking?

Reply:
> I preserved it in the first commit: [4b87045#diff-21d70609ca9dc507b2d33f4a68d89bca3cd727191ffe3a6319b63c36e750deb2R215](https://github.com/bazelbuild/bazel/commit/4b870459f32b93177184446aed74705e1f2dccac#diff-21d70609ca9dc507b2d33f4a68d89bca3cd727191ffe3a6319b63c36e750deb2R215)
> 
> It's quite similar to scanning for `.runfiles` in the scripts own file path, but slightly more precise since we just know the suffix we need to strip.

Yes, lets just go with that.

> 1. Finding the Python runfiles root: I don't yet see why this should be difficult

It's difficult because (1) the precise layout of the Python side of runfiles is an implementation detail, and (2) Python itself doesn't have an equivalent concept as Bazel's runfiles directory.

The proposed usage of `__file__` assumes the layout is `<root>/<repo>/<path>`, but a layout such as e.g. `<root>/site-packages/<libname>/<path>` works just as well as far as Python code is concerned.

[The `__file__` attribute is also optional](https://docs.python.org/3/reference/import.html#file__), so whether it works also depends on how the runfiles library was imported, and the particular details of any custom importer.

Stated another way: this runfiles library is only usable for a particular implementation of the Python rules and resulting layout of the files in runfiles. While this may sound innocuous at face value, as a language rule owner, its very unappealing because it means anyone using this runfiles library breaks when we want to change the language rule implementation. Except a user doesn't see "the runfiles library was relying on implementation details", they just see "the language rules changed and broke me".

> 2. Using that information to derive the Bazel repository containing the caller:
>    This is where we invoke a lot of magic that may break in a myriad of ways: It requires investigating the call stack, assumes that the caller is a file, assumes that it lies in the runfiles root, and so on.
>    Luckily for us, if this magic procedure breaks from a particular location, it's not too bad: It just means that users can't look up their Bazel context _from that particular location_. Assuming they want to use runfiles, they will likely find some more well-behaved place in their program where the lookup is successful and then pass the repo name on to code that can then call `Rlocation` without any further magic involved.

I consider the "they can just pick another location to call their code" assumption a weak one. It assumes the person who picks the path is also the one that calls the runfiles APIs. This assumption breaks down because people like to build libraries. For example, a Django plugin that inherently understood runfiles. Such a case has a user specifying paths, while the actual runfiles API call is done by e.g. django code.